AMD/ROCm OCP Micro-scaling Format (mx-fp8/mx-fp4) Support

petrex opened the pull request:
- This pull request introduces support for the [OCP Micro-scaling (MX) format](https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf), with a focus on compatibility with AMD **ROCm 7.0** and the **gfx950** architecture.

  This PR also establishes the foundation for enabling MX-FPX features in [TorchAO](https://github.com/pytorch/ao/issues/2229) on the AMD platform.

- Validation (**ROCm 7.0** + **gfx950** required):

  `111 relevant tests passing.`
  
  > PYTORCH_TEST_WITH_ROCM=1 python test/test_matmul_cuda.py -k test_blockwise -v
  
  Co-author: @jagadish-amd —  Thank you for the efforts leading validation on gfx950 with ROCm 7.0.

-----------------------------------

This pull request introduces support for new scalar types and scaling methods, particularly for ROCm 7.0 and gfx950, and refines testing for these features. Key changes include adding constraints for matrix dimensions, enabling block-wise scaling, and updating tests to accommodate new data types.

### Support for new scalar types and scaling methods:
* [`aten/src/ATen/cuda/CUDABlas.cpp`](diffhunk://#diff-74fcb26047c1df4024105d36ce22a36b77cf8cc93c28631d743e639b3d6066aeR1876-R1885): Added constraints for matrix dimensions when using `Float8_e8m0fnu` with block-wise scaling, ensuring dimensions are multiples of 32. Updated compatibility checks to support ROCm 7.0 for `Float8_e8m0fnu` and `Float8_e4m3fn`. [[1]](diffhunk://#diff-74fcb26047c1df4024105d36ce22a36b77cf8cc93c28631d743e639b3d6066aeR1876-R1885) [[2]](diffhunk://#diff-74fcb26047c1df4024105d36ce22a36b77cf8cc93c28631d743e639b3d6066aeL1913-R1934)

* [`aten/src/ATen/native/cuda/Blas.cpp`](diffhunk://#diff-e8a569efee1e650172f120a0fdcda024fe3e4703a4ee3336425c8f685af6b3abR1276-R1290): Introduced block-wise scaling for `Float8_e8m0fnu`, with checks for ROCm 7.0 and GPU architecture `gfx950`. Added validation for supported scalar types and matrix dimensions. [[1]](diffhunk://#diff-e8a569efee1e650172f120a0fdcda024fe3e4703a4ee3336425c8f685af6b3abR1276-R1290) [[2]](diffhunk://#diff-e8a569efee1e650172f120a0fdcda024fe3e4703a4ee3336425c8f685af6b3abR1349-R1364)

### Updates to scalar type mappings:
* [`aten/src/ATen/cuda/CUDADataType.h`](diffhunk://#diff-9188bb13b1a49f459141f5f9b875593d1c5ce2beb5ad711fdbaf5bc7089ec015L93-R93): Extended scalar type mappings to support `Float4_e2m1fn_x2` for ROCm 7.0.

* [`aten/src/ATen/cuda/tunable/GemmHipblaslt.h`](diffhunk://#diff-bfa1a3b5d4bef1892bf50338775f3b0fd8cd31fc1868148f3968b98aefb68e3fR88-R96): Added a constexpr mapping for `Float4_e2m1fn_x2` based on ROCm version.

### Enhancements to testing(@jagadish-amd): 
* [`test/test_matmul_cuda.py`](diffhunk://#diff-3f31c52b48cfddf8f4617d809f7695b2e4a1c78656f8c4b5143a4b45d01fcf23R765-R766): Updated tests to include new scalar types (`Float4_e2m1fn_x2`) and recipes (`mxfp4`). Added logic to handle different scaling recipes and validate compatibility with ROCm and CUDA versions. [[1]](diffhunk://#diff-3f31c52b48cfddf8f4617d809f7695b2e4a1c78656f8c4b5143a4b45d01fcf23R765-R766) [[2]](diffhunk://#diff-3f31c52b48cfddf8f4617d809f7695b2e4a1c78656f8c4b5143a4b45d01fcf23L1331-R1356) F592e669L1353R1472)

These changes improve compatibility with newer hardware and software versions, enhance functionality for matrix operations, and ensure robust testing for the added features.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

Commit:
Enhance support for Float8 and Float4 data types in scaled_gemm and related functions. Update error messages to reflect ROCm 6.5 compatibility. Add HIP data type mapping for Float4_e2m1fn_x2. Ensure proper version checks for ROCm in CUDA operations.

pruthvistony commented:
@petrex ,
Can we enable the UTs which test FP4 for _scaled_gemm()

petrex commented:
> @petrex , Can we enable the UTs which test FP4 for _scaled_gemm()

fp4 support requires gfx950. any plan to deploy that in CI?

Commit:
Update support for Float8 and Float4 data types in CUDA operations. Adjust error messages to reflect ROCm 7.0 compatibility and ensure matrix dimension checks for block-wise scaling. Modify HIP data type mappings and version checks for enhanced compatibility.

Commit:
Add support for MX GEMM in CUDA operations

Implement a new function to evaluate platform support for MX GEMM based on ROCm version and CUDA availability. Update the PLATFORM_SUPPORTS_MX_GEMM variable to utilize this new evaluation method, enhancing compatibility checks for GPU architectures.

Commit:
Add HIP data type mapping for CUDA_R_4F_E2M1

This update introduces a new mapping for the CUDA_R_4F_E2M1 data type to its corresponding HIP type, enhancing compatibility in CUDA to HIP conversions.

Commit:
Enhance data scaling functions for Float4 and Float8 types in CUDA tests

This update introduces a new power of 2 constant for `torch.float4_e2m1fn_x2` and modifies the `data_to_mx_scale` function to support different scaling recipes. Additionally, the test cases are updated to handle the new `mxfp4` recipe, ensuring proper scaling and compatibility checks for both `mxfp8` and `mxfp4` formats in matrix multiplication tests.

Commit:
Enhance bias checks in scaled matrix multiplication for CUDA

This update improves the validation of bias tensor types in the `_scaled_mm_out_cuda` function. It adds additional checks to ensure that the bias tensor is compatible with the output tensor's data type, specifically enforcing that biases must be BFloat16 or Half when the output is of certain types. This change aims to prevent runtime errors and improve the robustness of matrix multiplication operations in CUDA.

Commit:
Update ROCm version checks for Float4 and Float8 types in CUDA

This commit modifies the ROCm version checks for the Float4_e2m1fn_x2, Float8_e5m2, and Float8_e4m3fn data types in the `_scaled_mm_out_cuda` function. The version requirements are updated to ensure compatibility with the appropriate ROCm versions, enhancing the robustness of matrix multiplication operations in CUDA.

drisspg commented on test/test_matmul_cuda.py:
```diff
@@ -1512,13 +1522,14 @@ def test_blockwise_mxfp8_nvfp4_numerics(self, test_case_name, fast_accum, mkn, r
-        A_scale = to_blocked(A_scale)
-        B_scale = to_blocked(B_scale)
+        if not torch.version.hip:
+            A_scale = to_blocked(A_scale)
```
The row,major layout of the scales is exactly whats expected for rocm? OOC Is this maximal perf?

petrex replied:
As far as I recall, all applicable tests on ROCm 7.0 are passing. There may be additional tests that aren’t currently supported in our setup(gfx950 && ROCm 7.0).

drisspg commented:
Is this code exercised in CI?

jagadish-amd commented:
> Is this code exercised in CI?
> 
> Update: just saw
> 
> > fp4 support requires gfx950. any plan to deploy that in CI?

right, not yet on CI (requires gfx950)

Commit:
Refactor data type handling in _scaled_mm_out_cuda for improved clarity

This commit updates the parameter names for data type handling in the _scaled_mm_out_cuda function, changing 'a_scale_type' and 'b_scale_type' to 'a_scale_dtype' and 'b_scale_dtype' respectively. This change enhances code readability and consistency in handling data types for matrix multiplication operations in CUDA.

Commit:
Enhance CUDABlas scaling support for ROCm and CUDA versions

Updated the get_scale_mode function to include support for ROCm version 7.0 and above alongside CUDA version 12.8. This change ensures compatibility with HIPBLASLT for ROCm and maintains the existing CUBLASLT functionality for CUDA.

Commit:
Update scaling choice in CUDA Blas implementation to BlockWise1x32

Modified the scaling choice condition in the _scaled_mm_out_cuda function to specify BlockWise1x32, ensuring compatibility with the latest GPU architectures and enhancing the precision of scaling operations.

Commit:
lint

Commit:
Refactor row-wise scaling condition in CUDABlas implementation

This commit introduces a boolean variable to simplify the condition checking for row-wise scaling in the scaled_gemm function. This change enhances code readability and maintains compatibility with existing scaling operations for ROCm and CUDA.

Commit:
Refactor scaling attribute handling in scaled_gemm function

This commit streamlines the handling of scaling attributes for different data types in the scaled_gemm function. It clarifies the conditions for Float8 data types and enhances compatibility checks for CUDA and ROCm versions. Additionally, it updates comments to reflect the differences in implementation between AMD/ROCm and NVIDIA, improving code readability and maintainability.

petrex committed:
Merge branch 'main' into rocm_mx_fp4

Commit:
Fix scaling choice condition in _scaled_mm_out_cuda for BlockWise1x32 support

Commit:
Add skip condition for Float8_e4m3fn in FP8Matmul tests on ROCm CI setup

petrex commented:
Test failures are irrelevant, lint fixed. for mi355 CI, refer to https://github.com/pytorch/pytorch/pull/160215

Commit:
Add skip condition for tests on ROCm platform in ReproTestsDevice class

Commit:
Fix formatting of PLATFORM_SUPPORTS_FP8 initialization in common_cuda.py

petrex commented:
@pytorchbot merge -f 'tests failures irrelevant(CPU based)'

weifengpy commented:
this seems to be causing torchtitan/torchao failures for H100, see https://github.com/pytorch/ao/issues/2843

drisspg commented:
@pytorchbot revert

drisspg commented:
hmm no clean revert, can you forward fix this 
@petrex 
https://github.com/petrex/pytorch/blob/73022c3d3262c9adb351a4a8c739b98ad3c5ff94/aten/src/ATen/cuda/CUDABlas.cpp#L1868-L1875

it should be good on 12.9 +

petrex commented:
@drisspg imho a patch is maybe easier than revert. See the discussion on ao thread?
